phlib/nativethread.c

```diff
@@ -811,32 +811,204 @@ NTSTATUS PhGetThreadLastStatusValue(
+                mtaUsageGlobals = CoGetMTAUsageInfo_I();
+            }
+        }
+
+        PhEndInitOnce(&initOnce);
+    }
+
+    if (!mtaUsageGlobals)
+        return STATUS_UNSUCCESSFUL;
+
+    if (MTAInits)
+    {
+        status = NtReadVirtualMemory(
+            ProcessHandle,
+            mtaUsageGlobals->MTAInits,
+            MTAInits,
+            sizeof(ULONG),
+            NULL
+            );
+
+        if (!NT_SUCCESS(status))
+            return status;
+    }
+
+    if (MTAIncInits)
+    {
+        status = NtReadVirtualMemory(
+            ProcessHandle,
+            mtaUsageGlobals->MTAIncInits,
+            MTAIncInits,
+            sizeof(ULONG),
+            NULL
+            );
+
+        if (!NT_SUCCESS(status))
+            return status;
+    }
+
+    return STATUS_SUCCESS;
+}
+
+/**
+ * Determines if a thread belongs to the main STA by checking if it owns a main STA window.
+ *
+ * \param[in] ThreadId The identifier of the thread.
+ *
+ * \return Whether the thread owns a main STA window.
+ */
+BOOLEAN PhIsThreadIdMainSTA(
+    _In_ HANDLE ThreadId
+    )
+{
+    static CONST UNICODE_STRING oleMainThreadWndClass = RTL_CONSTANT_STRING(L"OleMainThreadWndClass");
+    static CONST UNICODE_STRING oleMainThreadWndName = RTL_CONSTANT_STRING(L"OleMainThreadWndName");
+    static PH_INITONCE initOnce = PH_INITONCE_INIT;
+    static HWND (WINAPI* NtUserFindWindowEx_I)(
+        _In_opt_ HWND hwndParent,
+        _In_opt_ HWND hwndChild,
+        _In_ PCUNICODE_STRING ClassName,
+        _In_ PCUNICODE_STRING WindowName,
+        _In_ ULONG Type
+        ) = NULL;
+    static ULONG_PTR(WINAPI* NtUserQueryWindow_I)(
+        _In_ HWND WindowHandle,
+        _In_ WINDOWINFOCLASS WindowInfo
+        ) = NULL;
+    HWND hwnd;
+
+    if (PhBeginInitOnce(&initOnce))
+    {
+        PVOID win32u;
+
+        if (win32u = PhGetLoaderEntryDllBaseZ(L"win32u.dll"))
+        {
+            NtUserFindWindowEx_I = PhGetDllBaseProcedureAddress(win32u, "NtUserFindWindowEx", 0);
+            NtUserQueryWindow_I = PhGetDllBaseProcedureAddress(win32u, "NtUserQueryWindow", 0);
+        }
+
+        PhEndInitOnce(&initOnce);
+    }
+
+    if (!NtUserFindWindowEx_I || !NtUserQueryWindow_I)
+        return FALSE;
+
+    hwnd = NULL;
+    do
+    {
+        // Find the next main STA window
+        hwnd = NtUserFindWindowEx_I(
+            HWND_MESSAGE,
+            hwnd,
+            &oleMainThreadWndClass,
+            &oleMainThreadWndName,
+            0
+            );
+
+        // Check if it belongs to the specified thread ID
+    } while (hwnd && NtUserQueryWindow_I(hwnd, WindowThread) != (ULONG_PTR)ThreadId);
+
+    return !!hwnd;
+}
+
+/**
+ * Determines if a thread belongs to the main STA by checking if it owns a main STA window.
+ *
+ * \param[in] ThreadHandle A handle to the thread. The handle must have THREAD_QUERY_LIMITED_INFORMATION access.
+ *
+ * \return Whether the thread owns a main STA window.
+ */
+BOOLEAN PhIsThreadMainSTA(
```

Comment:
It might be better to have a single function and enum to check against?

https://learn.microsoft.com/en-us/windows/win32/cossdk/threading-model-attribute

```
    Main Thread Apartment
    Single Thread Apartment
    Free Thread Apartment
    Neutral Apartment
    Any Apartment
````


PhIsThreadApartment(_In_ HANDLE ThreadHandle, enum value to compare?) similar to `PhGetThreadApartment`

Reply:
It's a helper function for telling apart main and non-main STA since we cannot do that via OLE TLS flags. I probably should've used a `Php` prefix and avoid putting it in the header. I think I'll just merge it into `PhGetThreadApartment` altogether.